File: game.py
"""GameEngine - 状态机驱动的游戏主循环"""
from __future__ import annotations
import random
import logging
from typing import List, Optional, Dict, Any

from models import (
    PlayerState, Role, PlayerConfig, Phase,
    GameLog, NightAction, DeathRecord, DeathReason,
)
from player import AIPlayer
from win_checker import check_win, game_over_info
import phases
import prompts

logger = logging.getLogger("ai-werewolf")

# 角色分配模板（9人局）
ROLE_TEMPLATE = [
    Role.WEREWOLF, Role.WEREWOLF, Role.WEREWOLF,
    Role.SEER,
    Role.WITCH,
    Role.HUNTER,
    Role.VILLAGER, Role.VILLAGER, Role.VILLAGER,
]


class GameEngine:
    """游戏引擎 - 管理整个游戏流程"""

    def __init__(self, player_configs: List[PlayerConfig],
                 seed: Optional[int] = None,
                 verbose: bool = False,
                 dry_run: bool = False,
                 llm_config: Optional[Dict] = None):
        self.verbose = verbose
        self.dry_run = dry_run
        self.seed = seed

        if seed is not None:
            random.seed(seed)

        llm_config = llm_config or {}
        max_retries = llm_config.get("max_retries", 3)
        temperature = llm_config.get("temperature", 0.7)
        max_tokens = llm_config.get("max_tokens", 512)

        # 创建玩家
        self.players: List[AIPlayer] = []
        for cfg in player_configs:
            state = PlayerState(
                player_id=cfg.player_id,
                name=cfg.name,
                model=cfg.model,
                role=Role.VILLAGER,  # 临时，之后分配
                personality=cfg.personality,
            )
            player = AIPlayer(
                state=state,
                max_retries=max_retries,
                temperature=temperature,
                max_tokens=max_tokens,
                dry_run=dry_run,
            )
            self.players.append(player)

        self.log = GameLog(players=[p.state for p in self.players], seed=seed)
        self.current_night: Optional[NightAction] = None
        self.seer_results: List[Dict[str, Any]] = []

    def setup(self) -> None:
        """分配角色"""
        roles = list(ROLE_TEMPLATE)
        random.shuffle(roles)

        for player, role in zip(self.players, roles):
            player.state.role = role
            # 女巫初始化药物
            if role == Role.WITCH:
                player.state.witch_has_antidote = True
                player.state.witch_has_poison = True
            # 猎人初始化
            if role == Role.HUNTER:
                player.state.hunter_can_shoot = True

        self.log.phase = Phase.SETUP

        if self.verbose:
            for p in self.players:
                logger.info(f"  角色分配: [{p.id}] {p.name} → {p.role.display_name}")

        # 发送身份提示
        self._send_role_reveals()

    def _send_role_reveals(self) -> None:
        """向每个玩家发送身份信息"""
        for player in self.players:
            teammates = None
            if player.role == Role.WEREWOLF:
                teammates = [
                    p.state for p in self.players
                    if p.role == Role.WEREWOLF and p.id != player.id
                ]
            prompt = prompts.role_reveal_prompt(player.state, teammates, player.state.personality)
            # 不需要 LLM 回复，仅记录
            player.state.speech_history.append({
                "type": "role_reveal",
                "content": prompt,
            })

    def run(self) -> GameLog:
        """运行整局游戏"""
        self.setup()

        max_days = 20  # 安全上限
        for day in range(1, max_days + 1):
            self.log.day = day

            if self.verbose:
                logger.info(f"\n{'='*50}")
                logger.info(f"  第 {day} 天 — 夜晚阶段")
                logger.info(f"{'='*50}")

            # ─── 夜晚 ───
            self.current_night = NightAction(day=day)

            phases.phase_wolf_kill(self)
            phases.phase_seer_check(self)
            phases.phase_witch_action(self)
            deaths = phases.phase_night_resolve(self)

            # 检查胜负
            winner, reason = check_win([p.state for p in self.players])
            if winner:
                self.log.winner = winner
                self.log.winner_reason = reason
                self._log_game_over()
                return self.log

            # ─── 白天 ───
            if self.verbose:
                logger.info(f"\n{'='*50}")
                logger.info(f"  第 {day} 天 — 白天阶段")
                logger.info(f"{'='*50}")

            deaths_info = phases.phase_day_announce(self, deaths)

            # 猎人开枪（夜晚死亡触发）
            if deaths:
                hunter_shot = phases.phase_hunter_shoot(self, deaths)
                if hunter_shot is not None:
                    # 猎人开枪后再次检查胜负
                    winner, reason = check_win([p.state for p in self.players])
                    if winner:
                        self.log.winner = winner
                        self.log.winner_reason = reason
                        self._log_game_over()
                        return self.log

            # 发言
            speeches = phases.phase_speech(self, day, deaths_info)

            # 投票
            vote_result = phases.phase_vote(self, day, speeches)

            # 投票放逐后的猎人开枪
            if vote_result.eliminated is not None:
                elim_player = self.get_player_by_id(vote_result.eliminated)
                if elim_player and elim_player.role == Role.HUNTER:
                    vote_death = DeathRecord(
                        player_id=vote_result.eliminated,
                        reason=DeathReason.VOTE_OUT,
                        day=day,
                    )
                    hunter_shot = phases.phase_hunter_shoot(self, [vote_death])
                    if hunter_shot is not None:
                        winner, reason = check_win([p.state for p in self.players])
                        if winner:
                            self.log.winner = winner
                            self.log.winner_reason = reason
                            self._log_game_over()
                            return self.log

            # 检查胜负
            winner, reason = check_win([p.state for p in self.players])
            if winner:
                self.log.winner = winner
                self.log.winner_reason = reason
                self._log_game_over()
                return self.log

        # 超过最大天数
        self.log.winner = "draw"
        self.log.winner_reason = f"游戏超过 {max_days} 天，判定平局。"
        return self.log

    def _log_game_over(self) -> None:
        self.log.phase = Phase.GAME_OVER
        info = game_over_info(
            [p.state for p in self.players],
            self.log.winner or "",
            self.log.winner_reason,
        )
        if self.verbose:
            logger.info(f"\n{info}")

    # ─── 辅助 ───

    def get_player_by_id(self, pid: int) -> Optional[PlayerState]:
        for p in self.players:
            if p.id == pid:
                return p.state
        return None

    def get_player_name(self, pid: int) -> str:
        p = self.get_player_by_id(pid)
        return p.name if p else f"未知({pid})"

File: lib/engine/game.ts
/**
 * GameEngine —— 状态机驱动的游戏主循环
 *
 * 对应原 Python 版 game.py:GameEngine，加：
 *   - async run()：LLM 调用天然异步
 *   - EventEmitter 注入：关键节点发事件，供 SSE / DB / 统计消费
 *   - recordLlmCall：每次 LLM 调用记录指标
 *   - seeded RNG（mulberry32）：dry-run 时结果可复现
 */
import type {
  GameLog,
  LlmUsage,
  NightAction,
  PlayerConfig,
  PlayerState,
  Role,
} from '../types';
import { createEmptyGameLog, createPlayerState, DEATH_REASON } from '../types';
import type { AIPlayerConfig } from '../llm/player';
import { AIPlayer } from '../llm/player';
import { EventEmitter, type EventType } from '../events';
import { checkWin, gameOverInfo } from './winChecker';
import { ROLE_TEMPLATES, validateLayout, type Layout } from './presets';
import * as prompts from '../prompts';
import * as phases from './phases';

export interface GameEngineOptions {
  playerConfigs: PlayerConfig[];
  layout?: Layout;
  seed?: number | null;
  verbose?: boolean;
  dryRun?: boolean;
  llmConfig?: AIPlayerConfig;
  emitter?: EventEmitter;
}

const PHASE_LABELS: Record<string, string> = {
  setup: '准备阶段',
  night_start: '夜幕降临',
  wolf_kill: '狼人行动',
  seer_check: '预言家查验',
  witch_save: '女巫决策',
  night_end: '夜晚结算',
  day_announce: '黎明到来',
  hunter_shoot: '猎人开枪',
  speech: '白天讨论',
  vote: '投票环节',
  vote_result: '投票结果',
  game_over: '游戏结束',
};

export class GameEngine {
  readonly verbose: boolean;
  readonly dryRun: boolean;
  readonly seed: number | null;
  readonly layout: Layout;
  readonly players: AIPlayer[] = [];
  readonly log: GameLog;
  readonly emitter: EventEmitter;
  currentNight: NightAction | null = null;
  seerResults: Array<{ id: number; name: string; isWolf: boolean }> = [];
  private rng: () => number;

  constructor(opts: GameEngineOptions) {
    this.verbose = opts.verbose ?? false;
    this.dryRun = opts.dryRun ?? false;
    this.seed = opts.seed ?? null;
    this.layout = opts.layout ?? '9p';
    this.emitter = opts.emitter ?? new EventEmitter();
    this.rng = createRng(this.seed);

    validateLayout(this.layout, opts.playerConfigs.length);

    const llmConfig = opts.llmConfig ?? {};
    for (const cfg of opts.playerConfigs) {
      const state = createPlayerState(cfg, 'villager'); // 临时角色，setup() 中分配
      const player = new AIPlayer(state, {
        ...llmConfig,
        dryRun: this.dryRun,
        verbose: this.verbose,
      });
      this.players.push(player);
    }

    this.log = createEmptyGameLog(
      this.players.map((p) => p.state),
      this.seed,
    );
  }

  setup(): void {
    const roles = [...ROLE_TEMPLATES[this.layout]];
    fisherYatesShuffle(roles, this.rng);

    for (const [i, player] of this.players.entries()) {
      const role = roles[i];
      player.state.role = role;
      // 守卫/女巫/猎人初始化状态
      if (role === 'witch') {
        player.state.witchHasAntidote = true;
        player.state.witchHasPoison = true;
      }
      if (role === 'hunter') {
        player.state.hunterCanShoot = true;
      }
    }

    this.log.phase = 'setup';

    this.emit(
      'game_start',
      {
        players: this.players.map((p) => ({
          playerId: p.id,
          name: p.name,
          model: p.state.model,
          role: p.role,
          personality: p.state.personality,
        })),
        layout: this.layout,
        seed: this.seed,
      },
      false,
    );

    if (this.verbose) {
      for (const p of this.players) {
        console.info(`  角色分配: [${p.id}] ${p.name} → ${p.role}`);
      }
    }

    // 给每个玩家发身份提示（不调 LLM，仅记录到历史）
    for (const player of this.players) {
      let teammates: PlayerState[] = [];
      if (player.role === 'werewolf') {
        teammates = this.players
          .filter((p) => p.role === 'werewolf' && p.id !== player.id)
          .map((p) => p.state);
      }
      const prompt = prompts.roleRevealPrompt(
        player.state,
        teammates,
        player.state.personality,
      );
      player.state.speechHistory.push({
        type: 'role_reveal',
        content: prompt,
      });
    }
  }

  async run(): Promise<GameLog> {
    this.setup();

    const maxDays = 20;
    for (let day = 1; day <= maxDays; day++) {
      this.log.day = day;

      if (this.verbose) {
        console.info(`\n${'='.repeat(50)}`);
        console.info(`  第 ${day} 天 — 夜晚阶段`);
        console.info('='.repeat(50));
      }

      // ─── 夜晚 ───
      this.currentNight = null;
      this.emitPhaseChange('night_start', day);

      await phases.phaseGuardProtect(this);
      this.emitPhaseChange('wolf_kill', day);
      await phases.phaseWolfKill(this);
      this.emitPhaseChange('seer_check', day);
      await phases.phaseSeerCheck(this);
      this.emitPhaseChange('witch_save', day);
      await phases.phaseWitchAction(this);
      this.emitPhaseChange('night_end', day);
      const deaths = phases.phaseNightResolve(this);

      // 胜负判定
      const win = checkWin(this.allStates());
      if (win.winner) {
        this.finishGame(win.winner, win.reason);
        return this.log;
      }

      // ─── 白天 ───
      if (this.verbose) {
        console.info(`\n${'='.repeat(50)}`);
        console.info(`  第 ${day} 天 — 白天阶段`);
        console.info('='.repeat(50));
      }

      this.emitPhaseChange('day_announce', day);
      const deathsInfo = phases.phaseDayAnnounce(this, deaths);

      // 猎人开枪（夜晚死亡触发）
      if (deaths.length) {
        this.emitPhaseChange('hunter_shoot', day);
        const hunterShot = await phases.phaseHunterShoot(this, deaths);
        if (hunterShot !== null) {
          const w2 = checkWin(this.allStates());
          if (w2.winner) {
            this.finishGame(w2.winner, w2.reason);
            return this.log;
          }
        }
      }

      // 发言
      this.emitPhaseChange('speech', day);
      const speeches = await phases.phaseSpeech(this, day, deathsInfo);

      // 投票
      this.emitPhaseChange('vote', day);
      const voteResult = await phases.phaseVote(this, day, speeches);

      // 被投票放逐若是猎人，开枪
      if (voteResult.eliminated !== null) {
        const elimPlayer = this.getPlayerStateById(voteResult.eliminated);
        if (elimPlayer && elimPlayer.role === 'hunter') {
          this.emitPhaseChange('hunter_shoot', day);
          const voteDeath = {
            playerId: voteResult.eliminated,
            reason: DEATH_REASON.VOTE_OUT,
            day,
            details: '被投票放逐',
          };
          const hunterShot = await phases.phaseHunterShoot(this, [voteDeath]);
          if (hunterShot !== null) {
            const w3 = checkWin(this.allStates());
            if (w3.winner) {
              this.finishGame(w3.winner, w3.reason);
              return this.log;
            }
          }
        }
      }

      // 胜负判定
      const w4 = checkWin(this.allStates());
      if (w4.winner) {
        this.finishGame(w4.winner, w4.reason);
        return this.log;
      }
    }

    // 超过最大天数
    this.finishGame('draw', `游戏超过 ${maxDays} 天，判定平局。`);
    return this.log;
  }

  private finishGame(winner: 'wolf' | 'good' | 'draw', reason: string): void {
    this.log.phase = 'game_over';
    this.log.winner = winner;
    this.log.winnerReason = reason;
    this.emitPhaseChange('game_over', this.log.day);
    this.emit(
      'game_over',
      {
        winner,
        reason,
        finalPlayers: this.allStates().map((p) => ({
          playerId: p.playerId,
          name: p.name,
          role: p.role,
          alive: p.alive,
        })),
      },
      false,
    );
    if (this.verbose) {
      console.info(gameOverInfo(this.allStates(), winner, reason));
    }
  }

  // ─── 事件 / 指标辅助 ──────────────────────────────────────

  emit(type: EventType, payload: Record<string, unknown>, isPrivate = false) {
    this.emitter.emit(type, this.log.day, this.log.phase, payload, isPrivate);
  }

  private emitPhaseChange(phase: string, day: number) {
    this.log.phase = phase as GameLog['phase'];
    this.emit(
      'phase_change',
      { phase, label: PHASE_LABELS[phase] ?? phase },
      false,
    );
  }

  recordLlmCall(player: AIPlayer, usage: LlmUsage | null) {
    if (!usage) return;
    this.emit(
      'llm_call',
      {
        playerId: player.id,
        playerName: player.name,
        model: player.state.model,
        usage,
      },
      true, // 统计事件标 private，UI 默认不显示
    );
  }

  // ─── 玩家/状态查询 ────────────────────────────────────────

  allStates(): PlayerState[] {
    return this.players.map((p) => p.state);
  }

  getPlayerStateById(pid: number): PlayerState | null {
    return this.players.find((p) => p.id === pid)?.state ?? null;
  }

  getPlayerName(pid: number): string {
    const p = this.getPlayerStateById(pid);
    return p ? p.name : `未知(${pid})`;
  }

  ensureCurrentNight(day: number): NightAction {
    if (!this.currentNight) {
      this.currentNight = {
        day,
        wolfTarget: null,
        guardProtected: null,
        seerTarget: null,
        seerResult: null,
        witchSaved: false,
        witchPoisoned: null,
        deaths: [],
      };
    }
    return this.currentNight;
  }

  nightHistorySummary(): prompts.NightHistoryEntry[] {
    return this.log.nights.map((n) => ({
      day: n.day,
      deaths: n.deaths.map((d) => ({ name: this.getPlayerName(d.playerId) })),
    }));
  }
}

// ─── RNG / shuffle ───────────────────────────────────────────

/**
 * Mulberry32 — 简单快速的 seeded PRNG。
 * 给定相同 seed，输出序列确定；不传 seed 则用 Math.random。
 */
function createRng(seed: number | null): () => number {
  if (seed === null) return Math.random;
  let s = seed >>> 0;
  return function () {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function fisherYatesShuffle<T>(arr: T[], rng: () => number): T[] {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}
